File: js/graph/DisjointSet.js
class DisjointSet {
    size;
    parent;

    constructor ( size ) {
        this.size = new Array(size).fill(0);
        this.parent = new Array(size).fill().map((v, idx) => idx);
    }

    find ( u ) {
        if ( this.parent[u] === u ) return u;
        return this.parent[u] = this.find(this.parent[u]);
    }

    union ( u, v ) {
        const par_u = this.find(u);
        const par_v = this.find(v);
        if ( par_u === par_v ) return;
        if ( this.size[par_u] > this.size[par_v] ) {
            this.parent[par_v] = par_u;
            this.size[par_u] += this.size[par_v];
        } else {
            this.parent[par_u] = par_v;
            this.size[par_v] += this.size[par_u];
        }
    }
}

File: java/graph/src/main/java/org/example/DisjointSet.java
package org.example;

public class DisjointSet {
    int[] size;
    int[] parent;
    public DisjointSet ( int n ) {
        this.size = new int[n + 1];
        this.parent = new int[n + 1];

        for ( int i = 0; i < this.size.length; i++ ) {
            this.size[i] = 1;
        }

        for ( int i = 0; i < this.parent.length; i++ ) {
            this.parent[i] = i;
        }
    }

    int find ( int u ) {
        if ( parent[u] == u ) return u;
        return parent[u] = find(parent[u]);
    }

    void union ( int u, int v ) {
        int ulp_u = find(u);
        int ulp_v = find(v);

        if ( ulp_u == ulp_v ) return;
        if ( size[ulp_u] < size[ulp_v] ) {
            parent[ulp_u] = ulp_v;
            size[ulp_v] += size[ulp_u];
        } else {
            parent[ulp_v] = ulp_u;
            size[ulp_u] += size[ulp_v];
        }
    }
}
